pkg/controller/storagecluster/cephcluster.go

```diff
@@ -366,24 +400,46 @@ func newStorageClassDeviceSets(sc *ocsv1.StorageCluster) []rook.StorageClassDevi
 
 		for i := 0; i < replica; i++ {
 			placement := rook.Placement{}
+			preparePlacement := rook.Placement{}
 
 			if noPlacement {
-				in := getPlacement(sc, "osd")
-				(&in).DeepCopyInto(&placement)
-
-				if len(topologyKeyValues) >= replica {
-					podAffinityTerms := placement.PodAntiAffinity.PreferredDuringSchedulingIgnoredDuringExecution
-					podAffinityTerms[0].PodAffinityTerm.TopologyKey = topologyKey
-
-					topologyIndex := i % len(topologyKeyValues)
-					nodeZoneSelector := corev1.NodeSelectorRequirement{
-						Key:      topologyKey,
-						Operator: corev1.NodeSelectorOpIn,
-						Values:   []string{topologyKeyValues[topologyIndex]},
+				if supportTSC {
+					in := getPlacement(sc, "osd-tsc")
+					(&in).DeepCopyInto(&placement)
+
+					if noPreparePlacement {
+						in := getPlacement(sc, "osd-prepare-tsc")
+						(&in).DeepCopyInto(&preparePlacement)
+
+						if len(topologyKeyValues) >= replica {
+							preparePlacement.TopologySpreadConstraints[0].TopologyKey = topologyKey
+						}
+					}
+				} else {
+					in := getPlacement(sc, "osd")
+					(&in).DeepCopyInto(&placement)
+
+					if noPreparePlacement {
+						in := getPlacement(sc, "osd-prepare")
+						(&in).DeepCopyInto(&preparePlacement)
+					}
+
+					if len(topologyKeyValues) >= replica {
```

Comment:
Can you explain why the topology keys are related to the replica? I would have thought we would always use a consistent topology key instead of using a different one depending on `i` as we iterate.

Reply:
As part of the old design, the topology key that we get from default.placements is `kubernetes.io/hostname`, this get replaced either by rack or zone in the `for loop`. the topology key doesn't depend on each `i `, it will be the same for every iteration. Only the values of node selector changes for each `i` like (us-east-2a, us-east-2b, us-east-2c)